File: python/aoc_2022_02.py
"""Advent of Code 2022 - Day 2."""

from doctest import testmod
from enum import Enum
from sys import stdin
from typing import Tuple


class Move(Enum):
    """Which move is being made."""

    ROCK = "ROCK"
    PAPER = "PAPER"
    SCISSORS = "SCISSORS"


beats = {Move.ROCK: Move.SCISSORS, Move.PAPER: Move.ROCK, Move.SCISSORS: Move.PAPER}
beaten = {q: p for p, q in beats.items()}
opponent_moves = {"A": Move.ROCK, "B": Move.PAPER, "C": Move.SCISSORS}
player_moves = {"X": Move.ROCK, "Y": Move.PAPER, "Z": Move.SCISSORS}


class Result(Enum):
    """Result of a game."""

    LOSE = "LOSE"
    DRAW = "DRAW"
    WIN = "WIN"


results = {"X": Result.LOSE, "Y": Result.DRAW, "Z": Result.WIN}


def read_moves(s: str) -> Tuple[Move, Move]:
    """Read a line as a pair of moves."""
    [opponent, player] = s.split()
    return (opponent_moves[opponent], player_moves[player])


def read_result(s: str) -> Tuple[Move, Result]:
    """Read a line as a move and a result."""
    [opponent, result] = s.split()
    return (opponent_moves[opponent], results[result])


def score_moves(moves: Tuple[Move, Move]) -> int:
    """Score a game given the two moves."""
    opponent, player = moves
    shape_score = 1 if player == Move.ROCK else 2 if player == Move.PAPER else 3
    if beats[opponent] == player:
        win_score = 0
    elif beats[player] == opponent:
        win_score = 6
    elif player == opponent:
        win_score = 3
    else:
        raise ValueError("Unknown outcome!")
    return win_score + shape_score


def score_result(game: Tuple[Move, Result]) -> int:
    """Score a game given the opponent moves and desired result."""
    opponent, result = game
    match result:
        case Result.LOSE:
            player = beats[opponent]
        case Result.WIN:
            player = beaten[opponent]
        case Result.DRAW:
            player = opponent
    return score_moves((opponent, player))


if __name__ == "__main__":
    testmod()
    lines = list(stdin)
    print(sum(score_moves(read_moves(line)) for line in lines))
    print(sum(score_result(read_result(line)) for line in lines))

File: rust/src/bin/aoc_2022_02.rs
// Advent of Code, 2022 day 2

use aoc_rust::stdin_lines;

#[derive(PartialEq)]
enum Outcome {
    Win,
    Draw,
    Lose,
}

impl Outcome {
    fn score(&self) -> i32 {
        match self {
            Outcome::Win => 6,
            Outcome::Draw => 3,
            Outcome::Lose => 0,
        }
    }
}

enum Shape {
    Rock,
    Paper,
    Scissors,
}

impl Shape {
    fn resolve(&self, other: &Self) -> Outcome {
        match (self, other) {
            (Shape::Rock, Shape::Scissors) => Outcome::Win,
            (Shape::Rock, Shape::Paper) => Outcome::Lose,
            (Shape::Paper, Shape::Rock) => Outcome::Win,
            (Shape::Paper, Shape::Scissors) => Outcome::Lose,
            (Shape::Scissors, Shape::Paper) => Outcome::Win,
            (Shape::Scissors, Shape::Rock) => Outcome::Lose,
            _ => Outcome::Draw,
        }
    }

    fn score(&self) -> i32 {
        match self {
            Shape::Rock => 1,
            Shape::Paper => 2,
            Shape::Scissors => 3,
        }
    }

    fn parse(c: char) -> Shape {
        match c {
            'A' => Shape::Rock,
            'B' => Shape::Paper,
            'C' => Shape::Scissors,
            _ => {
                panic!("Unknown shape");
            }
        }
    }

    fn next(&self) -> Shape {
        match self {
            Shape::Rock => Shape::Paper,
            Shape::Paper => Shape::Scissors,
            Shape::Scissors => Shape::Rock,
        }
    }
}

enum Response {
    X,
    Y,
    Z,
}

impl Response {
    fn parse(c: char) -> Response {
        match c {
            'X' => Response::X,
            'Y' => Response::Y,
            'Z' => Response::Z,
            _ => {
                panic!("Unknown response");
            }
        }
    }
    fn to_shape(&self) -> Shape {
        match self {
            Response::X => Shape::Rock,
            Response::Y => Shape::Paper,
            Response::Z => Shape::Scissors,
        }
    }
    fn to_outcome(&self) -> Outcome {
        match self {
            Response::X => Outcome::Lose,
            Response::Y => Outcome::Draw,
            Response::Z => Outcome::Win,
        }
    }
}

struct Game {
    opponent: Shape,
    response: Response,
}

impl Game {
    fn parse(s: &str) -> Game {
        assert!(s.len() == 3);
        Game {
            opponent: Shape::parse(s.chars().nth(0).expect("missing shape")),
            response: Response::parse(s.chars().nth(2).expect("missing response")),
        }
    }
}

fn score_a(g: &Game) -> i32 {
    let player = g.response.to_shape();
    let outcome = player.resolve(&g.opponent);
    outcome.score() + player.score()
}

fn score_b(g: &Game) -> i32 {
    let outcome = g.response.to_outcome();
    let mut player = Shape::Rock;
    while player.resolve(&g.opponent) != outcome {
        player = player.next()
    }
    outcome.score() + player.score()
}

fn main() {
    let games: Vec<Game> = stdin_lines().map(|s| Game::parse(&s)).collect();

    let part_a: i32 = games.iter().map(score_a).sum();
    let part_b: i32 = games.iter().map(score_b).sum();

    println!("{}", part_a);
    println!("{}", part_b);
}
